atc/lidar/scanner.go

```diff
@@ -0,0 +1,134 @@
+package lidar
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/concourse/atc/creds"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/pkg/errors"
+)
+
+func NewScanner(
+	logger lager.Logger,
+	checkFactory db.CheckFactory,
+	secrets creds.Secrets,
+	defaultCheckTimeout time.Duration,
+	defaultCheckInterval time.Duration,
+) *scanner {
+	return &scanner{
+		logger:               logger,
+		checkFactory:         checkFactory,
+		secrets:              secrets,
+		defaultCheckTimeout:  defaultCheckTimeout,
+		defaultCheckInterval: defaultCheckInterval,
+	}
+}
+
+type scanner struct {
+	logger lager.Logger
+
+	checkFactory         db.CheckFactory
+	secrets              creds.Secrets
+	defaultCheckTimeout  time.Duration
+	defaultCheckInterval time.Duration
+}
+
+func (s *scanner) Run(ctx context.Context) error {
+
+	lock, acquired, err := s.checkFactory.AcquireScanningLock(s.logger)
```

Comment:
More of a question, will having a scanning lock make it so that only one ATC can run scans on all resources at a time? And this makes it so that we can avoid having locks on the resource config?

Reply:
The scanning lock only guards against multiple atcs creating checks. We still need the resource config locks to guard against multiple atcs running the actual check.